File: 3sem/4lab/Banks/Exceptions/InvalidTransactionOperationException.cs
namespace Banks.Exceptions;

public class InvalidTransactionOperationException : BanksDomainException
{
    private InvalidTransactionOperationException(string? message)
        : base(message) { }

    public static InvalidTransactionOperationException OnTransactionToSameAccount()
        => new InvalidTransactionOperationException($"Unable to make transaction to the same bank account");

    public static InvalidTransactionOperationException OnIncorrectExecuting()
        => new InvalidTransactionOperationException($"Executed can be only transactions with status 'Created'");

    public static InvalidTransactionOperationException OnIncorrectCanceling()
        => new InvalidTransactionOperationException($"Canceled can be only transactions with status 'Executed'");

    public static InvalidTransactionOperationException OnInvalidTransaction()
        => new InvalidTransactionOperationException($"Unable to make or cancel transaction");
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/Exceptions/InvalidTransactionOperationException.java
package banksRemake.domain.exceptions;

public class InvalidTransactionOperationException extends BanksRemakeException {
    private InvalidTransactionOperationException(String message) {
        super(message);
    }

    public static InvalidTransactionOperationException onTransactionToSameAccount() {
        return new InvalidTransactionOperationException("Unable to make transaction to the same bank account");
    }

    public static InvalidTransactionOperationException onIncorrectExecuting() {
        return new InvalidTransactionOperationException("Executed can be only transactions with status 'Created'");
    }

    public static InvalidTransactionOperationException onIncorrectCanceling() {
        return new InvalidTransactionOperationException("Canceled can be only transactions with status 'Executed'");
    }

    public static InvalidTransactionOperationException onInvalidTransaction() {
        return new InvalidTransactionOperationException("Unable to make or cancel transaction");
    }
}
